File: backend/app/utils/course_text.py
"""Shared course-name normalization helpers for display and search."""

import re

from sqlalchemy import func


COURSE_PARENTHESIS_HALF_TO_FULL = str.maketrans({"(": "（", ")": "）"})
COURSE_PARENTHESIS_FULL_TO_HALF = str.maketrans({"（": "(", "）": ")"})


def format_course_display_name(value: str | None) -> str:
    if not value:
        return ""
    return str(value).translate(COURSE_PARENTHESIS_FULL_TO_HALF).strip()


def normalize_course_search_text(value: str | None) -> str:
    if value is None:
        return ""
    normalized = str(value).translate(COURSE_PARENTHESIS_FULL_TO_HALF)
    normalized = normalized.strip().lower()
    normalized = re.sub(r"[\s\u3000]+", "", normalized)
    normalized = normalized.replace("(", "").replace(")", "")
    return normalized


def _normalize_course_search_expression(value):
    expression = func.lower(func.trim(value))
    expression = func.replace(expression, "（", "(")
    expression = func.replace(expression, "）", ")")
    expression = func.replace(expression, "\u3000", "")
    expression = func.replace(expression, " ", "")
    expression = func.replace(expression, "\t", "")
    expression = func.replace(expression, "\n", "")
    expression = func.replace(expression, "(", "")
    expression = func.replace(expression, ")", "")
    return expression


def normalized_course_text_expr(*values):
    if not values:
        return func.lower("")

    normalized_values = [
        _normalize_course_search_expression(func.coalesce(func.nullif(func.trim(value), ""), ""))
        for value in values
    ]
    return func.coalesce(*normalized_values, "")

File: frontend/src/utils/courseText.js
const FULL_TO_HALF_PARENTHESIS = {
  '（': '(',
  '）': ')',
}

const fullToHalfRegex = new RegExp(`[${Object.keys(FULL_TO_HALF_PARENTHESIS).join('')}]`, 'g')

export const formatCourseDisplayName = (value) => {
  if (value == null) return ''
  return String(value)
    .replace(fullToHalfRegex, (char) => FULL_TO_HALF_PARENTHESIS[char])
    .trim()
}

export const normalizeCourseSearchText = (value) => {
  if (value == null) return ''
  return String(value)
    .replace(fullToHalfRegex, (char) => FULL_TO_HALF_PARENTHESIS[char])
    .replace(/\s+/g, '')
    .replace(/[()（）]/g, '')
    .toLowerCase()
    .trim()
}
